File: server/go/background/clash_poller.go
package background

import (
	"context"
	"encoding/json"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/luvxinc/vpn/server/models"
	"github.com/luvxinc/vpn/server/store"
)

// ClashPoller polls the sing-box Clash API every 10 seconds and writes traffic
// data to the access_log table.
type ClashPoller struct {
	DB         *store.DB
	RDB        *store.Redis
	ClashURL   string
	httpClient *http.Client
	// prevStats tracks last-seen upload/download per connection ID for delta calc.
	prevStats map[string]connStats
}

type connStats struct {
	upload   int64
	download int64
}

func NewClashPoller(db *store.DB, rdb *store.Redis, clashURL string) *ClashPoller {
	return &ClashPoller{
		DB:         db,
		RDB:        rdb,
		ClashURL:   clashURL,
		httpClient: &http.Client{Timeout: 5 * time.Second},
		prevStats:  make(map[string]connStats),
	}
}

// Run polls forever, stopping when ctx is cancelled.
func (p *ClashPoller) Run(ctx context.Context) {
	for {
		select {
		case <-ctx.Done():
			return
		case <-time.After(10 * time.Second):
		}
		if err := p.PollOnce(ctx); err != nil {
			slog.Debug("clash poller error (will retry)", "err", err)
		}
	}
}

// PollOnce fetches /connections from the Clash API and upserts access_log.
// Exported for testing.
func (p *ClashPoller) PollOnce(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.ClashURL+"/connections", nil)
	if err != nil {
		return err
	}
	resp, err := p.httpClient.Do(req)
	if err != nil {
		return nil // sing-box not running locally — skip silently
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var data struct {
		Connections []map[string]json.RawMessage `json:"connections"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return err
	}

	activeIDs := make(map[string]struct{}, len(data.Connections))

	for _, conn := range data.Connections {
		connIDRaw, ok := conn["id"]
		if !ok {
			continue
		}
		var connID string
		json.Unmarshal(connIDRaw, &connID)
		activeIDs[connID] = struct{}{}

		vlessUUID := extractVlessUUID(conn)
		if vlessUUID == "" {
			continue
		}

		sessionRaw, err := p.RDB.GetVlessMap(ctx, vlessUUID)
		if err != nil || sessionRaw == "" {
			continue
		}
		var info models.SessionInfo
		if err := json.Unmarshal([]byte(sessionRaw), &info); err != nil {
			continue
		}

		host := extractHost(conn)
		if host == "" {
			continue
		}

		var uploadTotal, downloadTotal int64
		if raw, ok := conn["upload"]; ok {
			json.Unmarshal(raw, &uploadTotal)
		}
		if raw, ok := conn["download"]; ok {
			json.Unmarshal(raw, &downloadTotal)
		}

		prev := p.prevStats[connID]
		uploadDelta := max64(0, uploadTotal-prev.upload)
		downloadDelta := max64(0, downloadTotal-prev.download)
		p.prevStats[connID] = connStats{upload: uploadTotal, download: downloadTotal}

		if uploadDelta == 0 && downloadDelta == 0 {
			continue
		}

		// Truncate to hour
		accessHour := time.Now().UTC().Truncate(time.Hour)
		if startRaw, ok := conn["start"]; ok {
			var startStr string
			if json.Unmarshal(startRaw, &startStr) == nil && startStr != "" {
				startStr = strings.Replace(startStr, "Z", "+00:00", 1)
				if t, err := time.Parse(time.RFC3339, startStr); err == nil {
					accessHour = t.UTC().Truncate(time.Hour)
				}
			}
		}

		userID, _ := uuid.Parse(info.UserID)
		sessionID, _ := uuid.Parse(info.SessionID)

		p.DB.UpsertAccessLog(ctx, userID, sessionID, host, accessHour, uploadDelta, downloadDelta)
		p.DB.UpdateSessionTraffic(ctx, sessionID, uploadDelta, downloadDelta)
	}

	// Prune stale delta entries
	for id := range p.prevStats {
		if _, ok := activeIDs[id]; !ok {
			delete(p.prevStats, id)
		}
	}

	return nil
}

// ExtractVlessUUID extracts the VLESS user UUID from connection metadata or chains.
// Exported for testing.
func ExtractVlessUUID(conn map[string]json.RawMessage) string {
	return extractVlessUUID(conn)
}

func extractVlessUUID(conn map[string]json.RawMessage) string {
	// Try metadata.inboundUser.uuid (sing-box 1.13+)
	if metaRaw, ok := conn["metadata"]; ok {
		var meta map[string]json.RawMessage
		if json.Unmarshal(metaRaw, &meta) == nil {
			if iuRaw, ok := meta["inboundUser"]; ok {
				var iu map[string]string
				if json.Unmarshal(iuRaw, &iu) == nil {
					if u, ok := iu["uuid"]; ok && u != "" {
						return u
					}
				}
			}
		}
	}
	// Fallback: scan chains for 36-char UUID-shaped strings
	if chainsRaw, ok := conn["chains"]; ok {
		var chains []string
		if json.Unmarshal(chainsRaw, &chains) == nil {
			for _, item := range chains {
				if len(item) == 36 && strings.Count(item, "-") == 4 {
					return item
				}
			}
		}
	}
	return ""
}

// extractHost returns the normalized host from connection metadata.
func extractHost(conn map[string]json.RawMessage) string {
	if metaRaw, ok := conn["metadata"]; ok {
		var meta map[string]string
		if json.Unmarshal(metaRaw, &meta) == nil {
			host := meta["host"]
			if host == "" {
				host = meta["destinationIP"]
			}
			if host != "" {
				return normalizeHost(host)
			}
		}
	}
	return ""
}

// NormalizeHost strips port from "host:port" and truncates to 253 chars. Exported for testing.
func NormalizeHost(host string) string { return normalizeHost(host) }

func normalizeHost(host string) string {
	if strings.HasPrefix(host, "[") {
		// IPv6 literal
	} else if idx := strings.LastIndex(host, ":"); idx >= 0 {
		host = host[:idx]
	}
	if len(host) > 253 {
		host = host[:253]
	}
	return host
}

func max64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

File: server/services/clash_poller.py
"""Background task: polls sing-box Clash API and writes access_log to PostgreSQL."""
import asyncio
import json
import logging
import uuid
from datetime import datetime, timezone

import httpx

log = logging.getLogger("weiai.clash_poller")

# Track the last-seen upload/download per connection ID to compute deltas
_prev_stats: dict[str, dict] = {}


async def run(app):
    """Runs forever. Poll Clash API every 10 seconds."""
    while True:
        try:
            await _poll_once(app)
        except asyncio.CancelledError:
            return
        except Exception as e:
            log.debug(f"Clash API poll error (will retry): {e}")
        await asyncio.sleep(10)


async def _poll_once(app):
    cfg = app.state.cfg
    clash_url = cfg["sing_box"]["clash_api_url"]
    db = app.state.db
    redis = app.state.redis

    async with httpx.AsyncClient(timeout=5.0) as client:
        try:
            resp = await client.get(f"{clash_url}/connections")
            resp.raise_for_status()
        except Exception:
            # sing-box not running locally during dev — skip silently
            return

        data = resp.json()
        connections = data.get("connections", [])
        active_ids = set()

        for conn in connections:
            conn_id = conn.get("id", "")
            active_ids.add(conn_id)

            # Extract vless UUID from the connection chains
            vless_uuid = _extract_vless_uuid(conn)
            if not vless_uuid:
                continue

            session_raw = await redis.get(f"vless_map:{vless_uuid}")
            if not session_raw:
                continue

            session_info = json.loads(session_raw)
            user_id = session_info["user_id"]
            session_id = session_info["session_id"]

            host = (
                conn.get("metadata", {}).get("host")
                or conn.get("metadata", {}).get("destinationIP")
                or "unknown"
            )
            if not host or host == "unknown":
                continue

            # Compute delta bytes since last poll
            upload_total = conn.get("upload", 0)
            download_total = conn.get("download", 0)
            prev = _prev_stats.get(conn_id, {"upload": 0, "download": 0})
            upload_delta = max(0, upload_total - prev["upload"])
            download_delta = max(0, download_total - prev["download"])
            _prev_stats[conn_id] = {"upload": upload_total, "download": download_total}

            if upload_delta == 0 and download_delta == 0:
                continue

            # Truncate timestamp to the current hour for aggregation
            start_str = conn.get("start", "")
            try:
                access_hour = datetime.fromisoformat(
                    start_str.replace("Z", "+00:00")
                ).replace(minute=0, second=0, microsecond=0)
            except Exception:
                access_hour = datetime.now(timezone.utc).replace(minute=0, second=0, microsecond=0)

            # Upsert into access_log (aggregate same host within the same hour)
            await db.execute(
                """INSERT INTO access_log
                     (user_id, session_id, host, access_hour, upload_bytes, download_bytes)
                   VALUES ($1, $2, $3, $4, $5, $6)
                   ON CONFLICT (session_id, host, access_hour)
                   DO UPDATE SET
                     request_count  = access_log.request_count + 1,
                     upload_bytes   = access_log.upload_bytes + EXCLUDED.upload_bytes,
                     download_bytes = access_log.download_bytes + EXCLUDED.download_bytes
                """,
                uuid.UUID(user_id),
                uuid.UUID(session_id),
                _normalize_host(host),
                access_hour,
                upload_delta,
                download_delta,
            )

            # Update running totals in sessions table
            await db.execute(
                """UPDATE sessions SET upload_bytes = upload_bytes + $1,
                                       download_bytes = download_bytes + $2
                   WHERE id = $3""",
                upload_delta, download_delta, uuid.UUID(session_id),
            )

        # Prune stale entries from our local delta tracker
        stale = [cid for cid in _prev_stats if cid not in active_ids]
        for cid in stale:
            del _prev_stats[cid]


def _extract_vless_uuid(conn: dict) -> str | None:
    """Extract the VLESS user UUID from the connection metadata or chains."""
    # sing-box Clash API puts the outbound chain in "chains"
    # The vless inbound user UUID appears in metadata.sourceIP correlation
    # In practice, we key off the outbound tag name since per-user UUIDs
    # are stored in Redis under vless_map:{uuid}. We iterate all known
    # active UUIDs to find a match — the Clash API doesn't expose the UUID
    # directly, but we can match via conn["id"] mapped to our Redis data.
    #
    # Better: sing-box 1.13+ includes "inboundUser" in connection metadata.
    metadata = conn.get("metadata", {})
    inbound_user = metadata.get("inboundUser", {})
    if isinstance(inbound_user, dict):
        return inbound_user.get("uuid")
    # Fallback: look for uuid-shaped string in chains
    for chain_item in conn.get("chains", []):
        if len(chain_item) == 36 and chain_item.count("-") == 4:
            return chain_item
    return None


def _normalize_host(host: str) -> str:
    """Strip port from host:port strings, truncate to 253 chars."""
    if ":" in host and not host.startswith("["):
        host = host.rsplit(":", 1)[0]
    return host[:253]
